lib/opte/src/engine/packet.rs

```diff
@@ -1249,32 +1225,47 @@ impl<T: Read + Pullup> Packet<FullParsed<T>> {
         T::Chunk: ByteSliceMut,
         T: Pullup,
     {
-        // If we know that no transform touched a field which features in
-        // an inner transport cksum (L4/L3 src/dst, most realistically),
-        // and no body transform occurred then we can exit early.
-        if !self.checksums_dirty() && !self.state.body_modified {
-            return;
-        }
-
-        // Flag to indicate if an IP header/ULP checksums were
-        // provided. If the checksum is zero, it's assumed heardware
-        // checksum offload is being used, and OPTE should not update
-        // the checksum.
-        let update_ip = self.state.meta.has_ip_csum();
-        let update_ulp = self.state.meta.has_ulp_csum();
-
-        // We expect that any body transform will necessarily invalidate
-        // the body_csum. Recompute from scratch.
-        if self.state.body_modified && (update_ip || update_ulp) {
+        // We expect that any body transform will necessarily invalidate the
+        // body_csum (and that of all headers). Recompute from scratch.
+        if self.state.body_modified {
             return self.compute_checksums();
         }
 
-        // Start by reusing the known checksum of the body.
-        let mut body_csum = self.body_csum().unwrap_or_default();
+        // If we know that no transform touched a field which features in an
+        // inner transport cksum (L4/L3 src/dst, most realistically) then we
+        // can exit early.
+        if !self.checksums_dirty() {
+            return;
+        }
 
-        // If a ULP exists, then compute and set its checksum.
-        if let (true, Some(ulp)) =
-            (update_ulp, &mut self.state.meta.headers.inner_ulp)
+        // In future we will want to factor in the offload information
+        // contained in the mblk_t. I would say that this gives us a perfect
+        // view on which checksums (other than UDP) have been omitted, but:
+        //
+        // * IP is in the unfortunate position where `HCK_IPV4_HDRCKSUM`
+        //   (requesting offload) and `HCK_IPV4_HDRCKSUM_OK` (the NIC verified
+        //   the checksum) are identical.
```

Comment:
This confused me until i looked up the illumos source. Maybe saying something like `HCK_IPV4_HDRCKSUM` and `HCK_IPV4_HDRCKSUM_OK` map to the same message block flag value would be clearer.